### ISSUE TITLE
Tensorflow is blocking CUDAmemcpyAsyn, and high CPU usage

### ISSUE THREAD

**Issue by MinghaoCheng** [tensorflow/tensorflow#40384]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution: Linux Ubuntu 18.04):
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): c++ 2.2.0
- Bazel version (if compiling from source): 2.0.0
- GCC/Compiler version (if compiling from source):7.5.0
- CUDA/cuDNN version: 10.1
- GPU model and memory: NVIDIA TitanV 12GB

**Describe the current behavior**
Background:
Hi, I'm using tensorflow c++ to do offline inference. Since I want to do it in realtime, I dag in and found some optimisation tricks. Perhaps the most prominent thing is, if and do the inference by calling session->infer(). (if I understand this correctly) Tensorflow would do copy data from RAM to VRAM, GPU calculation, copy result back to RAM. I found that the GPU utilisation is low because the GPU is waiting the data being fed. So I'm trying to do async data tranfer and calculation. (This could double the performance of my model since the GPU is only used by 40%, without data transfer, it can take up to 90%)

I found an example of creating GPU resident tensor in the source code of Tensorflow and I used it:

TensorShape shape = TensorShape({this->batchSize, this->input_height, this->input_width, this->input_depth});
    PlatformGpuId platform_gpu_id(0);

    GPUMemAllocator *sub_allocatorA =
        new GPUMemAllocator(
            GpuIdUtil::ExecutorForPlatformGpuId(platform_gpu_id).ValueOrDie(),
            platform_gpu_id, false /*use_unified_memory*/, {}, {});
    GPUBFCAllocator *allocatorA = new GPUBFCAllocator(sub_allocatorA, shape.num_elements() * sizeof(uint8), "GPU_0_bfc");
    this->BufferA = new Tensor(allocatorA, tensorflow::DT_UINT8, shape);

and I feed data using:

cudaMemcpyAsync(dst, p, this->batchSize * this->input_depth * this->input_height * this->input_width, cudaMemcpyHostToDevice, *CUDAstream);

Issue 1:
I implemented three threads, one is for feeding data, one is for inference, and one is for data fetching.

For some reason, session->run() would block the cudaMemcpyAsync() in which the cudaMemcpyAsync() is pending until session->run() ends.
This does not happen all the time, but for the majority of the time.

Issue 2:
As I mentioned, I'm using GPU for inference, but tensorflow would occupy all my CPU (44 cores 100%) when it's inferring.
I tried the code below, but it still creates roughly 100 threads and eat all the CPU. This is not happening with your official python wheels. (with the same saved model, there are still lots of threads created, but it is basically not using CPU). I can make sure the calculation is done by GPU, (I opened both htop and nvtop at the same time)

    auto options = SessionOptions();
    options.config.mutable_gpu_options()->set_allow_growth(true);
    tensorflow::ConfigProto &config = options.config;
    config.set_inter_op_parallelism_threads(10);
    config.set_intra_op_parallelism_threads(10);
    config.set_use_per_session_threads(false);  

**Describe the expected behavior**

expected 1:

Async memory transfer without being blocked.

expected 2:

basically same CPU usage with python when running model on GPU.

**Standalone code to reproduce the issue**
If you need it, I can push them on to github

**Other info / logs**
No other important info/logs, I'm working remotely using VNP&SSH during this hard time.

**Comment by sanjoy** [comment]
> If you need it, I can push them on to github

That would be great, please upload the code and post a link here if it isn't too much trouble.

**Comment by MinghaoCheng** [comment]
@sanjoy 
Thanks for your reply,
here is the link to the repo:
https://github.com/MinghaoCheng/tensorflowCC_infer_aync

**Comment by MinghaoCheng** [comment]
@sanjoy Hi,
Is there any update?
Thanks

**Comment by MinghaoCheng** [comment]
I think I worked it out. It is mainly openCV to blame, not tensorflow.
Thanks a lot.
